packages/SwingSet/src/kernel/liveSlots.js

```diff
@@ -461,6 +463,9 @@ function build(
       valToSlot.set(val, slot);
       slotToVal.set(slot, new WeakRef(val));
       if (type === 'object') {
+        // Set.delete() metering seems unaffected by presence/absence, but it
+        // doesn't matter anyway because deadSet.add only happens when
+        // finializers run, which happens deterministically
```

Comment:
Seems to me that this comment is not quite right.  Finalizers run non-deterministically -- that's why we use `deadSet` to protect ourselves.  It's the processing of `deadSet` that's run deterministically.

Reply:
By "finalizers run deterministically" I mean that, given the way `xsnap.c` uses XS, finalizer callbacks only run during the `gcAndFinalize()` window. XS only polls the registries during `fxEndJob()`, and we only call that after the promise queue is empty (https://github.com/Agoric/agoric-sdk/blob/cc06a29777a0f45f25c8f44b288c6dd87606df0a/packages/xsnap/src/xsnap.c#L1108):

```c
void fxRunLoop(txMachine* the)
{
	c_timeval tv;
	txNumber when;
	txJob* job;
	txJob** address;
	for (;;) {
		while (the->promiseJobs) {
			while (the->promiseJobs) {
				the->promiseJobs = 0;
				fxRunPromiseJobs(the);
			}
			// give finalizers a chance to run after the promise queue is empty
			fxEndJob(the);
			// if that added to the promise queue, start again
		}
		// at this point the promise queue is empty
 // .. the rest of the code checks for timer jobs
```

That's not mandatory: I think the JS spec allows engines to run the finalizer callbacks any time they want (as long as they happen in their own turn). The tricky part, for us, is metering. If the finalizer *could* happen at any arbitrary time, then it would get metered. The only thing a finalizer does is to examine the weakref and add non-revived vrefs to the deadset, which doesn't cause e.g. immediate syscalls, but even that would cause metering sensitivity to the state of the weakref:

```js
  function finalizeDroppedImport(vref) {
    const wr = slotToVal.get(vref);
    if (wr && !wr.deref()) {
      // we're in the COLLECTED state, or FINALIZED after a re-introduction
      deadSet.add(vref);
      slotToVal.delete(vref);
      // console.log(`-- adding ${vref} to deadSet`);
    }
  }
```

So our "metering is insensitive to GC timing" story already depends upon finalizers *not* running during the userspace phase.

If we were somehow *not* depending upon that, and finalizers could run at any old time, then a `vref` might or might not be in the `deadSet` when this `convertValToSlot` needed to delete it. We delete it because this might be a re-introduction of the vref: imagine a `dispatch.deliver()` crank which introduces `o-4`, then drops it, then performs a device invocation whose return value re-introduces `o-4`. If the finalizer could run in the middle of the crank, we'd have `o-4` in our `deadSet` when the `D()` return value was unserialized and `convertValToSlot('o-4')` creates a new `Presence`, but `o-4` is still in the `deadSet`. (Now that I think about it, we don't strictly need to delete it from `deadSet`.. `processDeadSet` will check the weakref anyways. But I think you and I decided that it would be tidier to delete it as soon as we know it's been re-introduced).

So in that case, `deadSet.delete('o-4')` would sometimes be deleting something, and sometimes not, depending upon this hypothetical non-deterministic finalizer callback having already run or not. *If* `.delete`'s metering usage were different in these two situations, then now the metering of `convertValToSlot` (i.e. `m.serialize()`) would become sensitive to GC state, which we don't want, because we want serialization to be metered, because (at least) it could invoke user-provided getters, which would allow user code to escape metering.

If that `deadSet.delete(slot)` were replaced with `if (deadSet.has(slot)) { deadSet.delete(slot); }`, that'd be obviously metering-affecting. But I wasn't sure about the built-in `deadSet.delete()`.. does it follow differently-metered code paths internally? I did an experiment and convinced myself that, at least right now, it does not: deleting an existing key and deleting a non-existing key both use the same metering.

So we're protected in two ways. 1: `deadSet.delete()`'s metering doesn't depend upon the presence/absence of the key, and 2: the key won't get added in the middle of a crank (the metered region) because that only happens in the finalizer callback, and our `xsnap.c` ensures that finalizers only happen at end-of-crank (when we call `gcAndFinalize`).

I'll update the comment to make it clear that the reason our finalizers run deterministically is because we wrote `xsnap.c` to make it that way, and that JS in general does not guarantee that in the slightest.